upload: Add command to reupload all custom emoji.

mateuszmandera opened the pull request:
This is for issue #19732 . The issue proposes adding this as a migration - but the issue with that is we rely on `get_emoji_url` and `upload_emoji_image` - and importing that kind of production code into is an issue in migration and we try to avoid it - usually we just copy-paste the logic into the migration file, which is probably not an option here. So for now, I wrote a management command, with the idea being that we have to add the instruction to run it in the release notes. 

Not ready to merge because:
1. I haven't tested it with S3 yet.
2. The issue proposes
>  `# Should we mark the first owner as the uploader of this custom emoji while we're at it? `

If we go with the approach of having a command for re-uploading, I think it'd make sense to do a separate migration for "mark the first owner as uploader" - rather than changing that within the management command.

mateuszmandera commented on zerver/management/commands/reupload_emoji.py:
```diff
@@ -0,0 +1,44 @@
+    def handle(self, *args: Any, **kwargs: str) -> None:
+        session = OutgoingSession(role="reupload_emoji", timeout=3, max_retries=3)
+
+        for realm_emoji in RealmEmoji.objects.order_by("id"):
```
Should we perhaps add an option to the command to run up from a specified id? ``--run-from-id 123``. It might be useful if the process fail halfway through, to not re-run it on all the emoji that already got processed.

timabbott replied:
I think that would probably end up being handy, and not hard to add, so I think that'd a good idea.

timabbott replied:
Maybe `--start-from=17`

mateuszmandera replied:
Added

timabbott commented on zerver/management/commands/reupload_emoji.py:
```diff
@@ -0,0 +1,44 @@
+from zerver.models import RealmEmoji, UserProfile
+
+
+class Command(sendtestemail.Command):
```
Why this choice of parent class?  Needs a comment.

mateuszmandera replied:
That was a mistake in a mistake i made in my vscode snippet for management commands, fixed.

timabbott commented on zerver/management/commands/reupload_emoji.py:
```diff
@@ -0,0 +1,44 @@
+            emoji_filename = realm_emoji.file_name
+            emoji_url = get_emoji_url(emoji_filename, realm_emoji.realm_id)
+            if emoji_url.startswith("/user_avatars"):
+                emoji_url = urljoin(realm_emoji.realm.uri, emoji_url)
```
Hmm, can we not use library code to do this URL computation?

mateuszmandera replied:
`/` or `/user_avatars` should be the same in terms of what happens, but changed to `/` in case that looks cleaner

timabbott replied:
Yeah, that does seem better.

timabbott commented on zerver/management/commands/reupload_emoji.py:
```diff
@@ -0,0 +1,44 @@
+            if emoji_url.startswith("/user_avatars"):
+                emoji_url = urljoin(realm_emoji.realm.uri, emoji_url)
+
+            print(f"Downloading {emoji_url}")
```
Should we be trying to download the `.original` files and working from them?  My vague recollection is that we've been saving them.

mateuszmandera replied:
Makes sense, changed this to use `.original` first and fallback to fetching the regular `emoji_url` if that fails for some reason

timabbott commented:
I think the management command option is reasonable; if we want to make it happen automatically, we can potentially run it from the upgrade tool.

mateuszmandera commented:
@timabbott Made the adjustments that popped up in review, added an entry to the changelog (not sure about the phrasing though) and verified this works with S3 after adding this line:
```
emoji_bytes_io.name = emoji_filename
```
(since the s3 codepath looks at the .name attribute)

So this should be ready for another review and merging.

mateuszmandera commented:
@timabbott Added a migration to set `.author` for emoji that don't have it as well. I think as a follow-up cleanup we should be able to set
```
    author: Optional["UserProfile"] = models.ForeignKey(
        "UserProfile",
        blank=True,
        null=True,
        on_delete=CASCADE,
    )
```

to non-NULL, get rid of `Optional` and clean up other code that's for handling the case where `.author = None`?

timabbott commented:
I think we can do that follow-up migration, though I think we might have cases where we don't want to do that -- e.g. if we move the `zulip` emoji to be a default custom emoji in the future.  I suppose we could have that be owned by Notification Bot or something if we need an owner?

mateuszmandera commented:
> I think we can do that follow-up migration, though I think we might have cases where we don't want to do that -- e.g. if we move the `zulip` emoji to be a default custom emoji in the future. I suppose we could have that be owned by Notification Bot or something if we need an owner?

Hmm, the issue description #19732 suggested the plan to perhaps set .author to `This would establish for the future the invariant that all custom emoji have an owner.`. So if we indeed want to establish that invariant (which seems nice to have), then yeah I think it'd be quite sensible to use `Notification bot` (or perhaps the first organization owner, like we're doing in this PR?) for the `author` for default custom emojis.

Fingel commented on zerver/management/commands/reupload_emoji.py:
```diff
@@ -0,0 +1,77 @@
+from zerver.models import RealmEmoji, UserProfile
+
+
+def get_emoji_file_content(session: OutgoingSession, emoji_url: str, emoji_id: int) -> bytes:
```
I am a bit concerned with the use of requests to fetch the emoji data, but it might be because I'm not familiar enough with the architecture. 

How do we know that the webserver will be running and reachable by this script? Custom installs might do weird stuff with proxies/nginx config. Also imagine you've deployed using Docker or a PAAS like heroku. Usually the shell you get (to run management commands) would be an instance of the container, not inside the container itself. In that case there might not be any webserver running and available locally at all.

timabbott replied:
I think this is a valid concern but I don't think there's a convenient alternative generic way to access the contents of uploaded files.

Probably this belongs in the documentation for using this tool -- you should run it while your Zulip server is operational.

mateuszmandera replied:
Yeah, I've added an explicit note that the server should be running to the upgrade notes and the help string of the command, good point. 

>  I don't think there's a convenient alternative generic way to access the contents of uploaded files.

@timabbott #20737 uses backend-specific code to fetch the files natively instead of this http request approach, but it gets more messy, so not sure if it's worth it

Fingel commented on zerver/management/commands/reupload_emoji.py:
```diff
@@ -0,0 +1,77 @@
+            emoji_file_content = get_emoji_file_content(session, emoji_url, realm_emoji.id)
+
+            emoji_bytes_io = io.BytesIO(emoji_file_content)
+            # The S3 backend code expects the .name attribute to be set.
```
I think this is actually a bug in https://github.com/mateuszmandera/zulip/blob/49e5100fd7551d2d10de25e41acea08da850e15f/zerver/lib/upload.py#L678 that is requiring you to add this.

See https://github.com/zulip/zulip/pull/20737/files#diff-8f1447ded7fcf78b852c341c06be7529a918aff84fa774959985a908ed524ab5R692

As mentioned in my review:

> 
>     The commit also fixes a potential bug at
>     zerver.lib.upload.S3UploadBackend:upload_emoji_image where the name
>     attribute was being accessed on the file object instead of using the
>     passsed in filename. Not all file objects in python are guaranteed to
>     have a name attribute.

mateuszmandera replied:
@Fingel  Ah! Yeah that's exactly the line, I was looking at it when working on this PR but didn't notice it was a bug and `emoji_file_name` should be used directly. I pushed that tweak as a separate prep commit, since we generally want to structure commits in this way (https://zulip.readthedocs.io/en/latest/contributing/version-control.html)

Fingel replied:
It did feel wrong to have that fix in the same commit as everything else in my PR. Nice job.

Fingel reviewed (commented):
Let me know what you think about using code to access the emoji data directly as opposed to making HTTP requests. Other than that I think the PR looks great.

timabbott commented on zerver/migrations/0375_set_realmemoji_author.py:
```diff
@@ -0,0 +1,34 @@
+        if realm_emoji.author_id is None:
+            user_profile = UserProfile.objects.filter(
+                realm_id=realm_emoji.realm_id, is_active=True, role=ROLE_REALM_OWNER
+            ).first()
```
Let's include an `order_by("id")` here -- we want to be specific about what "first" means.

timabbott replied:
Yeah, it's more a style thing -- if you care about the ordering, you should write it.

timabbott commented on zerver/management/commands/reupload_emoji.py:
```diff
@@ -0,0 +1,78 @@
+                # We only use the user_profile to get the realm
+                user_profile = UserProfile.objects.filter(
+                    realm=realm_emoji.realm, is_active=True, role=UserProfile.ROLE_REALM_OWNER
+                ).first()
```
Same comment on selecting the user as above.

mateuszmandera commented:
@timabbott I've switched this to the proposed deferred_worker-based approach, can you review?

I've also deleted the mention of anything  from Upgrade Notes, since I think this approach should mean no special action needs to be taken. Although I do somewhat wonder if there isn't *some* hypothetical room for complications, if `deferred_worker` starts processing these events while the server is down and thus it can't fetch the emoji via HTTP requests :thinking:

timabbott commented:
That sounds right, merged, thanks @mateuszmandera!

timabbott commented:
(Sadly `main` is failing due to PGroonga issues)
